app/liquid_tags/liquid_tag_base.rb

```diff
@@ -3,11 +3,54 @@ def self.script
     ""
   end
 
+  def validate_data(parsed_context)
+    source = parsed_context.partial_options[:source]
+    validate_contexts(source)
+
+    user = parsed_context.partial_options[:user]
+    validate_user_permissions(user)
+  end
+
   def finalize_html(input)
     input.gsub(/ {2,}/, "").
       gsub(/\n/m, " ").
       gsub(/>\n{1,}</m, "><").
       strip.
       html_safe
   end
+
+  private
+
+  def validate_contexts(source)
+    raise LiquidTagError, "No source found" unless source
+
+    is_valid_source = self.class::VALID_CONTEXTS.include? source.class_name
+    invalid_source_error_msg = "Invalid context. This liquid tag can only be used in #{self.class::VALID_CONTEXTS.join(', ')}."
+
+    raise LiquidTagError, invalid_source_error_msg unless is_valid_source
+  end
+
+  def validate_user_permissions(user)
+    raise LiquidTagError, "No user found" unless user
+
+    raise LiquidTagError, "User is not permitted to use this liquid tag" unless user_permitted_to_use_liquid_tag?(user)
+  end
+
+  def user_permitted_to_use_liquid_tag?(user)
+    self.class::VALID_ROLES.any? { |valid_role| user_has_valid_role?(user, valid_role) }
+  end
+
+  def user_has_valid_role?(user, valid_role)
+    # Single resource roles require 2 arguments - the role and the resource
+    if valid_role.is_a? Array
+      role, resource = valid_role
+
+      user.has_role? role, resource
+    else
+      user.has_role? valid_role
+    end
+  end
```

Comment:
When a role is a [single resource role](https://docs.dev.to/internal/), we need to define the role _and_ the resource. This method handles both. So, for example, `VALID_ROLES` might be something like:
```ruby
VALID_ROLES = [
  :super_admin,
  [:codeland_organizer, UserSubscriptionTag],
].freeze
```

So we handle both. And a liquid tag can have multiple valid roles.

Reply:
Should we have a LiquidTag Policy class for this? Might help contain the logic a bit in a single place rather than having the liquid tag worry about it.

Reply:
I agree with mstruve, it's a bit odd to have the role logic here in the liquid tag but maybe it can't moved up because of reasons related to parsing?

Reply:
Hmmm...I think I'm misunderstanding. Can you both clarify what you mean?

It's in a single location because all `LiquidTags` inherit from this base class. Or do you mean contain the piece where we are specifying roles in a single location? The roles will change for each liquid tag. So maybe put that "mapping" for each tag in a single location?

Reply:
Store all of the logic for determining if a liquid tag can be used or not in a Policy rather than here. https://github.com/thepracticaldev/dev.to/tree/master/app/policies since that is our convention for controlling access for most things

Reply:
If it's possible to put these permissions-related things into a specific `Policy` class, I agree that this would be preferable here.

Reply:
If you really can't use a policy class here for whatever reason, maybe we can at least remove the conditional logic:

```ruby
def user_has_valid_role?(user, valid_role)
  # Splat array for single resource roles
  user.has_role?(*Array(valid_role))
end
```